File: tests/verify/migration.py
"""An older, protocol-incompatible session server: fux explains, never kills it on its own, and
offers the operator an interactive choice. A fake old manager (answering FUXCTL1) stands in for
0.2.x; everything lives in a disposable runtime directory."""
import fcntl
import json
import os
import pty
import re
import select
import signal
import socket
import struct
import subprocess
import sys
import tempfile
import termios
import time
from pathlib import Path

FUX = str(Path(sys.argv[1]).resolve())
ESCAPES = re.compile(rb'\x1b\][^\x07\x1b]*(?:\x07|\x1b\\\\)|\x1b\[[0-9;?<>=]*[ -/]*[@-~]|\x1b[()][A-Za-z0-9]|\x1b[=>]')


def until(function, description, timeout=10):
    end = time.monotonic() + timeout
    while time.monotonic() < end:
        value = function()
        if value:
            return value
        time.sleep(.05)
    raise AssertionError(description)


def fake_old_server(runtime, ready):
    """Child process: a manager socket that speaks FUXCTL1 and exits cleanly on SIGTERM."""
    manager = runtime / 'fux/manager.sock'
    listener = socket.socket(socket.AF_UNIX)
    listener.bind(str(manager))
    os.chmod(manager, 0o600)
    listener.listen(8)

    def stop(*_):
        try:
            manager.unlink()
        finally:
            os._exit(0)
    signal.signal(signal.SIGTERM, stop)
    os.write(ready, b'1')
    os.close(ready)
    while True:
        try:
            stream, _ = listener.accept()
        except InterruptedError:
            continue
        with stream:
            stream.settimeout(2)
            try:
                stream.recv(8)
                stream.sendall(b'FUXCTL1\n')
            except OSError:
                pass


class Terminal:
    """fux running on a pty, with raw output accumulated for string checks."""

    def __init__(self, env):
        self.raw = b''
        self.status = None
        self.pid, self.fd = pty.fork()
        if self.pid == 0:
            fcntl.ioctl(0, termios.TIOCSWINSZ, struct.pack('HHHH', 24, 80, 0, 0))
            os.execve(FUX, [FUX], env)

    def pump(self, seconds=.05):
        end = time.monotonic() + seconds
        while time.monotonic() < end:
            ready, _, _ = select.select([self.fd], [], [], max(0, end - time.monotonic()))
            if not ready:
                continue
            try:
                data = os.read(self.fd, 65536)
            except OSError:
                return
            if not data:
                return
            self.raw += data

    def text(self):
        """Raw output with escape sequences removed; painted cells stay in order."""
        return ESCAPES.sub(b'', self.raw)

    def wait_for(self, needle, description, timeout=10):
        end = time.monotonic() + timeout
        while time.monotonic() < end:
            self.pump()
            if needle in self.text():
                return
            if self.exited():
                break
        raise AssertionError(f'{description}\n--- output ---\n{self.raw.decode(errors="replace")}')

    def send(self, data):
        os.write(self.fd, data)

    def drain(self):
        """Reads whatever fux has written so far: a pty buffer left full blocks the viewer in its
        frame write, and a blocked viewer cannot see the keys sent afterwards."""
        while True:
            ready, _, _ = select.select([self.fd], [], [], 0)
            if not ready:
                return
            try:
                data = os.read(self.fd, 65536)
            except OSError:
                return
            if not data:
                return
            self.raw += data

    def exited(self):
        self.drain()
        if self.status is None:
            child, status = os.waitpid(self.pid, os.WNOHANG)
            if child:
                self.status = os.waitstatus_to_exitcode(status)
        return self.status is not None

    def close(self):
        if not self.exited():
            os.kill(self.pid, signal.SIGKILL)
            os.waitpid(self.pid, 0)
        try:
            os.close(self.fd)
        except OSError:
            pass


with tempfile.TemporaryDirectory(prefix='fmig-', dir='/tmp') as directory:
    root = Path(directory)
    (root / 'fux/workspaces').mkdir(parents=True)
    os.chmod(root / 'fux', 0o700)
    os.chmod(root / 'fux/workspaces', 0o700)
    config = root / 'config/fux'
    config.mkdir(parents=True)
    (config / 'config.toml').write_text(
        'default-command = { argv = ["/bin/sh", "-c", "printf NEW_SERVER_PANE; exec cat"] }\n')
    env = os.environ.copy()
    env.update(HOME=directory, XDG_RUNTIME_DIR=directory, XDG_CONFIG_HOME=str(root / 'config'),
               XDG_STATE_HOME=str(root / 'state'), TERM='xterm-256color', SHELL='/bin/sh')

    read_end, write_end = os.pipe()
    old = os.fork()
    if old == 0:
        os.close(read_end)
        fake_old_server(root, write_end)
    os.close(write_end)
    assert os.read(read_end, 1) == b'1'
    os.close(read_end)
    (root / 'fux/workspaces/default.json').write_text(json.dumps({
        'name': 'default', 'pid': old, 'instance_nonce': 'old', 'socket_path': str(root / 'fux/default.attach.sock'),
        'protocol': 2}))

    def old_alive():
        try:
            os.kill(old, 0)
        except ProcessLookupError:
            return False
        return os.waitpid(old, os.WNOHANG)[0] == 0

    terminal = None
    try:
        # Without a terminal: explain and leave the old server alone.
        result = subprocess.run([FUX], env=env, stdin=subprocess.DEVNULL, capture_output=True, timeout=20)
        assert result.returncode != 0
        assert b'older protocol' in result.stderr and b'default (pid' in result.stderr, result.stderr
        assert b'XDG_RUNTIME_DIR' in result.stderr
        assert old_alive(), 'non-interactive fux must not touch the old server'

        # Interactive: quit keeps it; a refused confirmation keeps it.
        terminal = Terminal(env)
        terminal.wait_for(b'[k/s/q]', 'dialog prompt missing')
        terminal.send(b'q\n')
        until(terminal.exited, 'fux did not exit after q')
        assert terminal.status != 0 and old_alive()
        terminal.close()
        terminal = Terminal(env)
        terminal.wait_for(b'[k/s/q]', 'dialog prompt missing (2)')
        terminal.send(b'k\n')
        terminal.wait_for(b'Type "stop"', 'confirmation prompt missing')
        terminal.send(b'no\n')
        until(terminal.exited, 'fux did not exit after a refused confirmation')
        assert old_alive(), 'refused confirmation must not stop the old server'
        terminal.close()

        # Interactive: stop, confirm, and the new server takes over.
        terminal = Terminal(env)
        terminal.wait_for(b'[k/s/q]', 'dialog prompt missing (3)')
        terminal.send(b'k\n')
        terminal.wait_for(b'Type "stop"', 'confirmation prompt missing (2)')
        terminal.send(b'stop\n')
        until(lambda: not old_alive(), 'old server was not stopped', timeout=15)
        terminal.wait_for('NEW_SERVER_PANE'.encode(), 'new server pane did not render', timeout=20)
        terminal.send(b'\x01d')
        until(terminal.exited, 'viewer did not detach')
        assert terminal.status == 0, terminal.status
        listing = subprocess.run([FUX, 'default', 'list'], env=env, capture_output=True, timeout=10, check=True)
        assert json.loads(listing.stdout)['result']['value']['workspaces'][0]['name'] == 'default'
        print('PASS incompatible-server dialog: explains, keeps the old server unless confirmed, then replaces it', flush=True)
    finally:
        if terminal is not None:
            terminal.close()
        if old_alive():
            os.kill(old, signal.SIGKILL)
            os.waitpid(old, 0)
        subprocess.run([FUX, 'workspace', 'kill', 'default'], env=env, capture_output=True, timeout=10, check=False)

File: tools/xtask/src/scenarios/migration.rs
//! Incompatible-manager dialog leaves an old server alive until explicit confirmation.
use crate::support::{
    local::{Root, stop_servers, until},
    process::{self, Guard},
    terminal::Terminal,
};
use anyhow::{Result, ensure};
use serde_json::{Value, json};
use std::{
    fs,
    io::{Read, Write},
    os::unix::{fs::PermissionsExt, net::UnixListener},
    path::Path,
    process::Stdio,
    sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
    },
    time::Duration,
};

pub(super) fn old_manager(root: &Path) -> Result<()> {
    let path = root.join("fux/manager.sock");
    let stopping = Arc::new(AtomicBool::new(false));
    signal_hook::flag::register(signal_hook::consts::SIGTERM, stopping.clone())?;
    let listener = UnixListener::bind(&path)?;
    fs::set_permissions(&path, fs::Permissions::from_mode(0o600))?;
    listener.set_nonblocking(true)?;
    fs::write(root.join("old-ready"), b"1")?;
    while !stopping.load(Ordering::Relaxed) {
        match listener.accept() {
            Ok((mut peer, _)) => {
                peer.set_nonblocking(false)?;
                peer.set_read_timeout(Some(Duration::from_secs(2)))?;
                peer.set_write_timeout(Some(Duration::from_secs(2)))?;
                let mut bytes = [0; 8];
                if peer.read(&mut bytes).is_ok() {
                    let _ = peer.write_all(b"FUXCTL1\n");
                }
            }
            Err(error) if error.kind() == std::io::ErrorKind::WouldBlock => {
                std::thread::sleep(Duration::from_millis(10))
            }
            Err(error) if error.kind() == std::io::ErrorKind::Interrupted => {}
            Err(error) => return Err(error.into()),
        }
    }
    fs::remove_file(path)?;
    Ok(())
}
pub(super) fn run(binary: &Path) -> Result<()> {
    let root = Root::new(
        "fmig-rs-",
        &[
            "/bin/sh".into(),
            "-c".into(),
            "printf NEW_SERVER_PANE; exec cat".into(),
        ],
    )?;
    fs::create_dir_all(root.path().join("fux/workspaces"))?;
    for path in ["fux", "fux/workspaces"] {
        fs::set_permissions(root.path().join(path), fs::Permissions::from_mode(0o700))?;
    }
    let mut old = Guard(
        root.command(&std::env::current_exe()?)
            .args(["fixture-worker", "old-manager"])
            .arg(root.path())
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()?,
    );
    until(Duration::from_secs(5), || {
        ensure!(old.0.try_wait()?.is_none(), "old manager exited");
        Ok(root.path().join("old-ready").exists().then_some(()))
    })?;
    fs::write(
        root.path().join("fux/workspaces/default.json"),
        serde_json::to_vec(
            &json!({"name":"default","pid":old.0.id(),"instance_nonce":"old","socket_path":root.path().join("fux/default.attach.sock"),"protocol":2}),
        )?,
    )?;
    let scenario = (|| -> Result<()> {
        let result = process::output(root.command(binary), Duration::from_secs(20), 1024 * 1024)?;
        let stderr = String::from_utf8_lossy(&result.stderr);
        ensure!(
            !result.status.success(),
            "noninteractive migration succeeded"
        );
        ensure!(
            stderr.contains("older protocol")
                && stderr.contains("default (pid")
                && stderr.contains("XDG_RUNTIME_DIR"),
            "missing noninteractive explanation: {stderr}"
        );
        ensure!(
            old.0.try_wait()?.is_none(),
            "noninteractive fux stopped old manager"
        );
        {
            let mut terminal = Terminal::start(&root, binary)?;
            terminal.wait_for("[k/s/q]", Duration::from_secs(10))?;
            terminal.send(b"q\n")?;
            ensure!(
                !terminal.wait(Duration::from_secs(10))?.success() && old.0.try_wait()?.is_none(),
                "quit stopped old manager or succeeded"
            );
        }
        {
            let mut terminal = Terminal::start(&root, binary)?;
            terminal.wait_for("[k/s/q]", Duration::from_secs(10))?;
            terminal.send(b"k\n")?;
            terminal.wait_for("Type \"stop\"", Duration::from_secs(10))?;
            terminal.send(b"no\n")?;
            terminal.wait(Duration::from_secs(10))?;
            ensure!(old.0.try_wait()?.is_none(), "refusal stopped old manager");
        }
        let mut terminal = Terminal::start(&root, binary)?;
        terminal.wait_for("[k/s/q]", Duration::from_secs(10))?;
        terminal.send(b"k\n")?;
        terminal.wait_for("Type \"stop\"", Duration::from_secs(10))?;
        terminal.send(b"stop\n")?;
        until(Duration::from_secs(15), || {
            terminal.pump()?;
            Ok(old.0.try_wait()?)
        })?;
        terminal.wait_for("NEW_SERVER_PANE", Duration::from_secs(20))?;
        terminal.send(b"\x01d")?;
        ensure!(
            terminal.wait(Duration::from_secs(10))?.success(),
            "detach failed"
        );
        let mut listing = root.command(binary);
        listing.args(["default", "list"]);
        let listing = process::output(listing, Duration::from_secs(10), 1024 * 1024)?;
        ensure!(
            listing.status.success()
                && serde_json::from_slice::<Value>(&listing.stdout)?["result"]["value"]["workspaces"]
                    [0]["name"]
                    == "default",
            "new workspace not listed"
        );
        Ok(())
    })();
    drop(old);
    let cleanup = stop_servers(root.path());
    let failures: Vec<_> = [scenario, cleanup]
        .into_iter()
        .filter_map(Result::err)
        .map(|error| format!("{error:#}"))
        .collect();
    ensure!(failures.is_empty(), "{}", failures.join("; "));
    println!(
        "PASS incompatible-server dialog: explains, keeps the old server unless confirmed, then replaces it"
    );
    Ok(())
}
